test/bootloader_test.rb

```diff
@@ -0,0 +1,146 @@
+      expect(Yast::BootCommon).to receive(:setLoaderType).with(Yast::BootCommon.getLoaderType(true))
+      subject.Import({})
+    end
+
+    it "acts like missing if \"loader_type\" value is empty" do
+      expect(Yast::BootCommon).to receive(:setLoaderType).with(Yast::BootCommon.getLoaderType(true))
+      subject.Import("loader_type" => "")
+    end
+
+    it "pass initrd specific map to initrd module" do
+      initrd_map = {
+        "list"     => ["nouveau", "nvidia"],
+        "settings" => {
+          "nouveau" => {
+            "debug" => "1"
+          }
+        }
+      }
+      expect(Yast::Initrd).to receive(:Import).with(initrd_map)
+
+      subject.Import("initrd" => initrd_map)
+    end
+
+    it "sets passed \"write_settings\" map" do
+      write_settings = { "key" => "value" }
+
+      subject.Import("write_settings" => write_settings)
+      expect(Yast::BootCommon.write_settings).to eq write_settings
+    end
+  end
+
+  describe ".ReadOrProposeIfNeeded" do
+    before do
+      allow(subject).to receive(:Propose)
+      allow(subject).to receive(:Read)
+      Yast::BootCommon.was_read = false
+      Yast::BootCommon.was_proposed = false
+
+      allow(subject).to receive(:UpdateConfiguration)
+    end
+
+    it "does nothing if already read" do
+      expect(subject).to_not receive(:Propose)
+      expect(subject).to_not receive(:Read)
+      Yast::BootCommon.was_read = true
+
+      subject.ReadOrProposeIfNeeded
+    end
+
+    it "does nothing if already proposed" do
+      expect(subject).to_not receive(:Propose)
+      expect(subject).to_not receive(:Read)
+      Yast::BootCommon.was_proposed = true
+
+      subject.ReadOrProposeIfNeeded
+    end
+
+    it "does nothing in config mode" do
+      expect(subject).to_not receive(:Propose)
+      expect(subject).to_not receive(:Read)
+      expect(Yast::Mode).to receive(:config).and_return(true)
+
+      subject.ReadOrProposeIfNeeded
+    end
+
+    it "propose configuration in initial stage except update mode" do
+      expect(subject).to receive(:Propose)
+      expect(subject).to_not receive(:Read)
+      expect(Yast::Stage).to receive(:initial).and_return(true)
+      expect(Yast::Mode).to receive(:update).and_return(false)
+
+      subject.ReadOrProposeIfNeeded
+    end
+
+    it "reads configuration in normal stage" do
+      expect(subject).to_not receive(:Propose)
+      expect(subject).to receive(:Read)
+      expect(Yast::Stage).to receive(:initial).and_return(false)
+      allow(Yast::Mode).to receive(:update).and_return(false)
+
+      subject.ReadOrProposeIfNeeded
+    end
+
+    it "reads configuration in update mode" do
+      expect(subject).to_not receive(:Propose)
+      expect(subject).to receive(:Read)
+      expect(Yast::Stage).to receive(:initial).and_return(true)
+      allow(Yast::Mode).to receive(:update).and_return(true)
+
+      subject.ReadOrProposeIfNeeded
+    end
+
+    it "calls .UpdateConfiguration in update mode" do
```

Comment:
Then I will move the ```allow(subject).to receive(:UpdateConfiguration)``` from the before block and will put it only in the  "reads configuration in update mode". This way it would be more obvious (for instance, I wouldn't have asked).